File: pages/refactorings/replace-global-state-with-parameter/02-shared-settings/After.hs
-- Settle transactions: refuse any that would overdraw past the
-- limit, and charge a fee on each one accepted.
module After where

import Data.List (foldl')

data Policy = Policy { limit :: Int, fee :: Int }

policy :: Policy
policy = Policy { limit = 30, fee = 2 }

applyTx :: Policy -> Int -> Int -> Int
applyTx policy balance tx =
  let next = balance + tx
  in if next < -limit policy then balance
     else next - fee policy

settle :: [Int] -> Int
settle = foldl' (applyTx policy) 0

File: pages/refactorings/replace-global-state-with-parameter/02-shared-settings/After.scala
// Settle transactions: refuse any that would overdraw past the
// limit, and charge a fee on each one accepted.
object After:
  case class Policy(limit: Int, fee: Int)
  val policy = Policy(limit = 30, fee = 2)

  def applyTx(policy: Policy, balance: Int, tx: Int): Int =
    val next = balance + tx
    if next < -policy.limit then balance
    else next - policy.fee

  def settle(txs: List[Int]): Int =
    txs.foldLeft(0)(applyTx(policy, _, _))
